tsl/src/compression/compression.c

```diff
@@ -2319,19 +2323,47 @@ decompress_batches(RowDecompressor *decompressor, ScanKeyData *scankeys, int num
 									&tmfd,
 									false);
 
-		if (result == TM_Updated || result == TM_Deleted)
-		{
-			if (IsolationUsesXactSnapshot())
-				ereport(ERROR,
-						(errcode(ERRCODE_T_R_SERIALIZATION_FAILURE),
-						 errmsg("could not serialize access due to concurrent update")));
-			return false;
-		}
-		if (result == TM_Invisible)
+		switch (result)
 		{
-			elog(ERROR, "attempted to lock invisible tuple");
-			return false;
+			case TM_Deleted:
+			{
+				if (IsolationUsesXactSnapshot())
+				{
+					ereport(ERROR,
+							(errcode(ERRCODE_T_R_SERIALIZATION_FAILURE),
+							 errmsg("could not serialize access due to concurrent update")));
+					table_endscan(heapScan);
+					return false;
+				}
+				continue;
+			}
+			break;
+			case TM_BeingModified:
```

Comment:
```suggestion
			/* TM_BeingModified cannot happen since we always wait for locks here
			 * TM_SelfModified should never happen
			 */
			case TM_BeingModified:
```